File: Code/gait_control/src/motion.cpp
#include <stdio.h>
#include <math.h>
#include <iostream>
#define TERMINAL_PARAMETER_BOOL(name, desc, def) \
    bool name = def;
#define TERMINAL_PARAMETER_FLOAT(name, desc, def) \
    float name = def;
#define TERMINAL_PARAMETER_DOUBLE(name, desc, def) \
    double name = def;
#define TERMINAL_PARAMETER_INT(name, desc, def) \
    int name = def;

    /*
#ifdef __EMSCRIPTEN__
#include <emscripten/bind.h>
#else
#include "imu.h"
#include "voltage.h"
#include "distance.h"
#endif
#ifdef RHOCK
#include <rhock/event.h>
#include <rhock/stream.h>
#endif
*/
#include "config.h"
#include "cubic.h"
#include "motion.h"
#include "kinematic.h"
#include "mapping.h"

// Angles for the legs motor
float l1[4], l2[4], l3[4];

// Extra angles
float a1[4], a2[4], a3[4];

// Extra x, y and z for each leg
static float ex[4], ey[4], ez[4];

float motion_get_motor(int idx)
{
    int c = (idx%3);
    switch (c) {
        case 0:
            return (l1[idx/3]);
            break;
        case 1:
            return (l2[idx/3]);
            break;
        case 2:
            return (l3[idx/3]);
            break;
    }

    return 0;
}

// Amplitude multiplier
#define AMPLITUDE 30

// Speed factor
TERMINAL_PARAMETER_FLOAT(freq, "Time factor gain", 2.0);

// Legs bacakward mode
TERMINAL_PARAMETER_BOOL(backLegs, "Legs backwards", false);

// Amplitude & altitude of the robot
TERMINAL_PARAMETER_FLOAT(alt, "Height of the steps", 15.0);

// Static position
TERMINAL_PARAMETER_FLOAT(r, "Robot size", 153.0);
TERMINAL_PARAMETER_FLOAT(h, "Robot height", -55.0);

// Direction vector
TERMINAL_PARAMETER_FLOAT(dx, "Dx", 0.0);
TERMINAL_PARAMETER_FLOAT(dy, "Dy", 0.0);

// Turning, in ° per step
TERMINAL_PARAMETER_FLOAT(turn, "Turn", 0.0);

// Crab
TERMINAL_PARAMETER_FLOAT(crab, "Crab", 0.0);

// Front delta h
TERMINAL_PARAMETER_FLOAT(frontH, "Front delta H", 0.0);

// Gait selector
TERMINAL_PARAMETER_FLOAT(gait, "Gait", 1);

// Support
float support = 0.5;

// Functions
Cubic rise;
Cubic step;

/**
 * Initializing functions
 */
void setup_functions()
{
    rise.clear();
    step.clear();

    step.addPoint(0, 0.5, -1/support);
    step.addPoint(support, -0.5, -1/support);
    step.addPoint(support+(1-support)/2, 0, 1);
    step.addPoint(1, 0.5, -1/support);

    rise.addPoint(0, 0, 0);
    rise.addPoint(support, 0, 0);
    rise.addPoint(support+(1-support)/2, 1, 0);
    rise.addPoint(1, 0, 0);
}


TERMINAL_PARAMETER_FLOAT(smoothBackLegs, "Smooth 180", 0.0);

// Extra values
float extra_h = 0;
float extra_r = 0;

// Is the robot moving?
bool motion_is_moving()
{
    return (fabs(dx)>0.5 || fabs(dy)>0.5 || fabs(turn)>5);
}

void motion_init()
{
    // Setting the mapping to 0
    remap(0);

    for (int i=0; i<4; i++) {
        ex[i] = 0;
        ey[i] = 0;
        ez[i] = 0;
        a1[i] = 0;
        a2[i] = 0;
        a3[i] = 0;
    }

    extra_h = 0;
    extra_r = 0;
    freq = 2.0;
}

void motion_tick(float t)
{
    /*
    if (!motors_enabled()) {
        return;
    }
    */
    // Setting up functions
    setup_functions();

    // Smoothing 180
    if (backLegs && smoothBackLegs < 1) {
        smoothBackLegs += 0.02;
    }
    if (!backLegs && smoothBackLegs > 0) {
        smoothBackLegs -= 0.02;
    }

    float crabRad;
    float phasesA[] = {0.0, 0.5, 1-1e-6, 0.5};
    float phasesB[] = {0.0, 0.5, 0.75, 0.25};

    for (int i=0; i<4; i++) {
        // Defining in which group of opposite legs this leg is
        bool group = ((i&1)==1);

        // This defines the phase of the gait
        float legPhase;

        // Defining gait
        legPhase = t + phasesA[i]*gait + phasesB[i]*(1-gait);

        // Leg target
        float x, y, z, a, b, c;

        // Computing the order in the referencial of the body
        float stepping = step.getMod(legPhase);

        // Add the radius to the leg, in the right direction
        float radius = (r+extra_r);

        // The leg position in the body frame
        float X = (cos(M_PI/4)*radius) * ((i==0||i==1) ? 1 : -1);
        float Y = (cos(M_PI/4)*radius) * ((i==0||i==3) ? 1 : -1);
        float X_ = X;
        float Y_ = Y;

        // Applying crab
        crabRad = DEG2RAD(crab) * (group ? 1 : -1);
        X = cos(crabRad)*X_ - sin(crabRad)*Y_;
        Y = sin(crabRad)*X_ + cos(crabRad)*Y_;

        // Extras
        X += ex[i];
        Y += ey[i];

        // Add dX and dY to the moving vector
        if (fabs(turn) > 0.5) {
            float turnRad = -DEG2RAD(turn);
            float theta = -stepping*turnRad;
            float l = sqrt(dx*dx+dy*dy)/turnRad;
            float r = atan2(dy, dx);
            float cr = cos(-r);
            float sr = sin(-r);

            X_ = X; Y_ = Y;
            X = X_*cr - Y_*sr;
            Y = X_*sr + Y_*cr;

            X_ = X; Y_ = Y;
            X = X_*cos(theta) - (Y_+l)*sin(theta);
            Y = X_*sin(theta) + (Y_+l)*cos(theta) - l;

            X_ = X; Y_ = Y;
            X = X_*cr - Y_*(-sr);
            Y = X_*(-sr) + Y_*cr;
        } else {
            X += stepping*dx;
            Y += stepping*dy;
        }
        if(i==0){
        	std::cout << "legPhase: " << legPhase << std::endl;
        	std::cout << "stepping: " << stepping << std::endl;
        	std::cout <<"idx: "<< i << "leg pos: " << X <<","<< Y << std::endl;
        }
		//std::cout <<"idx: "<< i << "leg pos: " << X <<","<< Y << std::endl;
        // Move to the leg frame
        float vx, vy;
        legFrame(X, Y, &vx, &vy, i, L0);

        // The robot is moving if there is dynamics parameters
        bool moving = motion_is_moving();

        // This is the x,y,z order in the referencial of the leg
        x = vx;
        y = vy;
        z = ez[i] + h - extra_h + (moving ? (rise.getMod(legPhase)*alt) : 0);
        if (i < 2) z += frontH;

        // Computing inverse kinematics
        if (computeIK(x, y, z, &a, &b, &c, L1, L2, backLegs ? L3_2 : L3_1)) {
            l1[i] = -SIGN_A*a + a1[i];
            l2[i] = -SIGN_B*b + a2[i];
            l3[i] = -SIGN_C*(c - 180*smoothBackLegs) + a3[i];
        }
    }
}

void motion_reset()
{
    motion_init();

}

void motion_set_f(float f_)
{
    freq = f_;
}

float motion_get_f()
{
    return freq;
}

void motion_set_h(float h_)
{
    extra_h = h_;
}

void motion_set_r(float r_)
{
    extra_r = r_;
}

void motion_set_x_speed(float x_speed)
{
    dx = ODOMETRY_TRANSLATION*x_speed/(2.0*freq);
}

void motion_set_y_speed(float y_speed)
{
    dy = ODOMETRY_TRANSLATION*y_speed/(2.0*freq);
}

void motion_set_turn_speed(float turn_speed)
{
    turn = ODOMETRY_ROTATION*turn_speed/(2.0*freq);
}

void motion_extra_x(int index, float x)
{
    if (index >= 4) {
        for (int k=0; k<4; k++) {
            ex[k] = x;
        }
    } else {
        ex[index] = x;
    }
}

void motion_extra_y(int index, float y)
{
    if (index >= 4) {
        for (int k=0; k<4; k++) {
            ey[k] = y;
        }
    } else {
        ey[index] = y;
    }
}

void motion_extra_z(int index, float z)
{
    if (index >= 4) {
        for (int k=0; k<4; k++) {
            ez[k] = z;
        }
    } else {
        ez[index] = z;
    }
}

void motion_extra_angle(int index, int motor, float angle)
{
    for (int k=0; k<4; k++) {
        if (k == index || index >= 4) {
            if (motor == 0) a1[k] = angle;
            if (motor == 1) a2[k] = angle;
            if (motor == 2) a3[k] = angle;
        }
    }
}

float motion_get_dx()
{
    return dx;
}

float motion_get_dy()
{
    return dy;
}

float motion_get_turn()
{
    return turn;
}

File: Code/metabot/Motion/motion.py
import numpy as np
import Kinematics.Kinematics_firmware as kin

class point:
    def __init__(self, x, y, yp):
        self.x = x
        self.y = y
        self.yp = yp
        
class polynom:
    def __init__(self, a, b, c, d):
        self.a = a
        self.b = b
        self.c = c
        self.d = d

class cubic:
    def __init__(self):
        self.nbpoints = 0
        self.points = list()
        self.polynoms = list()
        
    def get_xmax(self):
        if self.nbpoints == 0:
            return 0.0
        else:
            return self.points[self.nbpoints-1].x
            
    def cal_polynom(self, pre_idx, curr_idx):
        if pre_idx >= self.nbpoints or pre_idx < 0:
            print("[ERROR] Previous Point Idx out of Range")
        if curr_idx >= self.nbpoints or curr_idx < 0:
            print("[ERROR] Current Point Idx out of Range")
        pre_point = self.points[pre_idx]
        curr_point = self.points[curr_idx]
        a = 2*pre_point.y + pre_point.yp - 2*curr_point.y + curr_point.yp
        b = -3*pre_point.y - 2*pre_point.yp + 3*curr_point.y - curr_point.yp
        c = pre_point.yp
        d = pre_point.y
        return a, b, c, d
        
    def add_point(self, x, y, yp):
        self.points.append(point(x, y, yp))
        self.nbpoints += 1
        if self.nbpoints > 1:
            a, b, c, d = self.cal_polynom(self.nbpoints-2, self.nbpoints-1)
            self.polynoms.append(polynom(a, b, c, d))
        
    def change_point(self, point_idx, y):
        if point_idx >= self.nbpoints:
            print("[ERROR] Point Idx out of Range")
            return
        else:
            x = self.points[point_idx].x
            yp = self.points[point_idx].yp
            self.points[point_idx] = point(x, y, yp)
            
        if point_idx < self.nbpoints - 1:
            a, b, c, d = self.cal_polynom(point_idx, point_idx+1)
            self.polynoms[point_idx] = polynom(a, b, c, d)
            
        if point_idx > 0:
            a, b, c, d = self.cal_polynom(point_idx-1, point_idx)
            self.polynoms[point_idx-1] = polynom(a, b, c, d)
        
    def get(self, x):
        for idx in range(self.nbpoints):
            if self.points[idx].x >= x:
                break
                
        if idx == 0 or idx == self.nbpoints:
            return 0.0
            
        A = self.points[idx-1]
        B = self.points[idx]
        f = self.polynoms[idx-1]
        x = (x-A.x)/(B.x-A.x)
        
        tx = x
        result = f.d
        result += f.c * tx
        tx *= x
        result += f.b * tx
        tx *= x
        result += f.a * tx
        
        return result
        
    def get_mod(self, x):
        max_x = self.get_xmax()
        if x < 0.0 or x > max_x:
            x = x % max_x
        return self.get(x)
             
class motion:
    def __init__(self):
        self.r = 153.0 # robot size
        self.h = -55.0 # robot height
        self.alt = 15.0 # robot step height
        self.rise_left = cubic()
        self.rise_right = cubic()
        self.step_left = cubic()
        self.step_right = cubic()
        self.phasesA = [0.0, 0.5, 1.0, 0.5]
        self.phasesB = [0.0, 0.5, 0.75, 0.25]
        self.gait = 1
        self.front_h_left = 0
        self.front_h_right = 0
        self.extra_r = 0
        self.extra_h = 0
        self.freq = 2.0
        self.dx = 0
        self.dy = 0
        self.pos = [[0.0,0.0,0.0],[0.0,0.0,0.0],[0.0,0.0,0.0],[0.0,0.0,0.0]]
        self.kin_solver = kin.Kinematics()
        
    def motion_tick(self, t):
        l1 = list()
        l2 = list()
        l3 = list()
        angle = np.arctan2(self.dy, self.dx) - np.pi/4 # angle to the first leg
        for leg_idx in range(4):
            leg_phase = t + self.phasesA[leg_idx] * self.gait + self.phasesB[leg_idx] * (1-self.gait)
            local_angle = angle + np.pi*leg_idx/2
            local_angle = local_angle - np.floor(local_angle/(2*np.pi))*2*np.pi
            is_left = local_angle < np.pi
            is_back = local_angle >= np.pi/2 and local_angle < np.pi*3/2
            if is_left:
                stepping = self.step_left.get_mod(leg_phase)
                rising = self.rise_left.get_mod(leg_phase)
            else:
                stepping = self.step_right.get_mod(leg_phase)
                rising = self.rise_right.get_mod(leg_phase)
            radius = self.r + self.extra_r
            
            x = np.cos(np.pi/4) * radius * (1 if leg_idx == 0 or leg_idx == 1 else -1)
            y = np.cos(np.pi/4) * radius * (1 if leg_idx == 0 or leg_idx == 3 else -1)

            x += stepping * self.dx
            y += stepping * self.dy
            
            #x, y = self.kin_solver.coordinate_transformation([X, Y], leg_idx)
            z = self.h - self.extra_h + rising * self.alt
            if is_back:
                #print("motion::leg_idx ",leg_idx)
                if is_left:
                    z -= self.front_h_left
                else:
                    z -= self.front_h_right
                
            self.pos[leg_idx] = [x, y, z]
            # compute IK
            L = self.kin_solver.IK([x, y, z], leg_idx)

            if L is None:
                print("[ERROR] Motion Error")
                return None
            else:
                l1.append(L[0])
                l2.append(L[1])
                l3.append(L[2])
                
        return l1, l2, l3
    
    def motion_set_f(self, f):
        self.freq = f
        
    def motion_set_h(self, h):
        self.extra_h = h
        
    def motion_set_r(self, r):
        self.extra_r = r
        
    def motion_set_front_h(self, h_left, h_right):
        self.front_h_left = h_left
        self.front_h_right = h_right
        
    def motion_set_x_speed(self, x_speed):
        self.dx = 1.2*x_speed/(2.0*self.freq)
        
    def motion_set_y_speed(self, y_speed):
        self.dy = 1.2*y_speed/(2.0*self.freq)

    def get_position(self):
        return self.pos
